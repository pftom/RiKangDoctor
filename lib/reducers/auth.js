Object.defineProperty(exports,"__esModule",{value:true});var _constants=require('redux-persist-immutable/constants');
var _immutable=require('immutable');var _immutable2=_interopRequireDefault(_immutable);


var _constants2=require('../constants/');










var _store=require('../store');function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}

var initialAuthState=_immutable2.default.Map({
token:null,
isLoggedIn:false,
isLoadingData:false,
loginError:false,
loginSuccess:false,
registerError:false});


var auth=function auth(){var state=arguments.length>0&&arguments[0]!==undefined?arguments[0]:initialAuthState;var action=arguments[1];
switch(action.type){


case _constants2.REGISTER:

return state.merge({
isLoadingData:true,
registerError:false});


case _constants2.REGISTER_ERROR:

return state.merge({
isLoadingData:false,
registerError:true});


case _constants2.LOGIN:

return state.merge({
isLoadingData:true,
loginError:false,
loginSuccess:false});


case _constants2.LOGIN_SUCCESS:


return state.merge({
isLoggedIn:true,
isLoadingData:false,
loginSuccess:true});


case _constants2.LOGIN_ERROR:


return state.merge({
isLoadingData:false,
loginError:true});


case _constants2.LOGOUT:

_store.persistor.purge();


return state.merge({
isLoggedIn:false,
token:null,
loginError:false,
loginSuccess:false});


case _constants2.CLEAR_ERROR:

return state.merge({
loginError:false,
registerError:false});


case _constants2.SET_TOKEN:var


payload=action.payload;
return state.merge({
token:payload});


case _constants2.CLEAR_TOKEN:


return state.merge({
token:null});


case _constants.REHYDRATE:var

_auth=action.payload.auth;
var token=_auth&&_auth.has('token')&&_auth.get('token');


var isLoggedIn=!!token?true:false;

return state.merge({
loginError:null,
registerError:false,
isLoggedIn:isLoggedIn,
token:token,
loginSuccess:false});


default:
return state;}

};exports.default=

auth;