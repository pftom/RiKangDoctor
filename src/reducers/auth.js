import { REHYDRATE } from 'redux-persist-immutable/constants';
import Immutable from 'immutable';

//import action constants
import { LOGIN_SUCCESS, LOGIN_ERROR, LOGOUT, LOGIN, CLEAR_TOKEN, SET_TOKEN } from '../constants/'
import { persistor } from '../store';

const initialAuthState = Immutable.Map({ 
  token: null,
  isLoggedIn: false,
  isLoadingData: false,
  loginError: false,
});

const auth = function auth(state = initialAuthState, action) {
  switch (action.type) {

    // later write the register logic

    case LOGIN: 
      //capture login action and show loading spinner
      return state.merge({
        isLoadingData: true,
        loginError: false,
      });

    case LOGIN_SUCCESS:

    //set the isLoggedIn
      return state.merge({
        isLoggedIn: true,
        isLoadingData: false,
      });

    case LOGIN_ERROR:

      //if login error, change the state 
      return state.merge({
        isLoadingData: false,
        loginError: true,
      })

    case LOGOUT:
      //clear all the persist data in the storage
      persistor.purge();

      //change the show screen
      return state.merge({
        isLoggedIn: false,
      })

    case SET_TOKEN: 

    //set the token for operate database later
      const { payload } = action;
      return state.merge({
        token: payload,
      });
      
    case CLEAR_TOKEN: 

      // 清除token
        return state.merge({
          token: null,
        });

    case REHYDRATE:
      //get persist data from redux-persist
      const { auth } = action.payload;
      const token = auth && auth.has('token') && auth.get('token');

      //use isLoggedIn show the initialScreen
      let isLoggedIn = !!token ? true : false;
      
      return state.merge({
        isLoggedIn,
        token
      })

    default:
      return state;
  }
}

export default auth;