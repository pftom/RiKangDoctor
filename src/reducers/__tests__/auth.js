import { Map } from 'immutable';

import auth from '../auth';

import {
  REGISTER,
  REGISTER_ERROR,
  REGISTER_SUCCESS,

  LOGIN,
  LOGIN_ERROR,
  LOGIN_SUCCESS,

  LOGOUT,

  CLEAR_STATE,
  CLEAR_TOKEN,

  
} from '../../constants/'

const registerAction = {
  type: REGISTER,
};

const registerErrorAction = {
  type: REGISTER_ERROR,
};

const registerSuccessAction = {
  type: REGISTER_SUCCESS,
};

const loginAction = {
  type: LOGIN,
};

const loginSuccessAction = {
  type: LOGIN_SUCCESS,
  token: "1234",
};

const loginErrorAction = {
  type: LOGIN_ERROR,
};

const logoutAction = {
  type: LOGOUT,
};

const clearStateAction = {
  type: CLEAR_STATE,
};

const clearTokenAction = {
  type: CLEAR_TOKEN,
};


const rehydrateHasToken = {
  type: 'REHYDRATE',
  payload: Map({
    auth: {
      token: '123',
    },
  }),
};

const rehydrateNoToken = {
  type: 'REHYDRATE',
  payload: Map({
    auth: {
      token: null,
    },
  }),
};

const initialAuthState = Map({ 
  token: null,
  isLoggedIn: false,
  isLoadingData: false,
  loginError: false,
  loginSuccess: false,
  registerError: false,
  registerSuccess: false,
});


test('the auth reducer work as well', () => {
  //test register reducer
  expect(auth(initialAuthState, registerAction)).toMatchSnapshot();
  expect(auth({...initialAuthState, isLoadingData: true}, registerErrorAction)).toMatchSnapshot();
  expect(auth({...initialAuthState, isLoadingData: true}, registerSuccessAction)).toMatchSnapshot();

  //test login reducer
  expect(auth(initialAuthState, loginAction)).toMatchSnapshot();
  expect(auth({...initialAuthState, isLoadingData: true}, loginSuccessAction)).toMatchSnapshot();
  //test login error reducer
  //no token
  expect(auth({...initialAuthState, isLoadingData: true}, loginErrorAction)).toMatchSnapshot();
  //have token
  expect(auth({...initialAuthState, token: "123", isLoadingData: true}, loginErrorAction)).toMatchSnapshot();
  //logout 
  expect(auth({...initialAuthState, token: "123"}, logoutAction)).toMatchSnapshot();
  //clear state
  expect(auth({...initialAuthState, loginError: true, clearStateAction: true }, logoutAction)).toMatchSnapshot();
  //clear token
  expect(auth({...initialAuthState, token: "123" }, clearTokenAction)).toMatchSnapshot();

  //rehydrate
  //has token
  expect(auth(initialAuthState, rehydrateHasToken)).toMatchSnapshot();
  //no token
  expect(auth(initialAuthState, rehydrateNoToken)).toMatchSnapshot();
})