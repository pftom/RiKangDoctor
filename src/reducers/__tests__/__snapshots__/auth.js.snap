// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the answer reducer work as well 1`] = `
Immutable.Map {
  answers: null,
  isLoadingData: true,
  loadingError: false,
  loadingSuccess: false,
}
`;

exports[`the answer reducer work as well 2`] = `
Immutable.Map {
  answers: Immutable.Map {
  },
  isLoadingData: false,
  loadingError: false,
  loadingSuccess: true,
}
`;

exports[`the answer reducer work as well 3`] = `
Immutable.Map {
  answers: null,
  isLoadingData: false,
  loadingError: true,
  loadingSuccess: false,
}
`;
